substrate-node/pallets/pallet-minting/src/lib.rs

```diff
@@ -0,0 +1,468 @@
+        node_report.last_period_uptime = node_report.counters.uptime;
+        // Refetch farming policy
+        node_report.counters.farming_policy = farming_policy_id;
+
+        let mut payable_periods = PayablePeriods::<T>::get(node_id);
+        payable_periods.push(node_report.counters);
+        PayablePeriods::<T>::insert(node_id, payable_periods);
+
+        Self::deposit_event(Event::NodePeriodEnded { node_id });
+        // Reset counters
+        node_report.counters = types::NodePeriodInformation::default();
+    }
+
+    // This method will calculate the minting rewards for all the periods that the node has accumulated
+    // Can only called by the farmer that owns the node
+    pub fn payout_node_periods(
+        account_id: &T::AccountId,
+        node_id: u32,
+    ) -> DispatchResultWithPostInfo {
+        let node = pallet_tfgrid::Nodes::<T>::get(node_id)
+            .ok_or(pallet_tfgrid::Error::<T>::NodeNotExists)?;
+
+        let twin_id = pallet_tfgrid::TwinIdByAccountID::<T>::get(&account_id)
+            .ok_or(pallet_tfgrid::Error::<T>::TwinNotExists)?;
+
+        let farm = pallet_tfgrid::Farms::<T>::get(node.farm_id)
+            .ok_or(pallet_tfgrid::Error::<T>::FarmNotExists)?;
+
+        ensure!(
+            twin_id == farm.twin_id,
+            Error::<T>::UnauthorizedToTriggerPayout
+        );
+
+        let mut payable_periods = PayablePeriods::<T>::get(node_id);
+        payable_periods = payable_periods
+            .into_iter()
+            .filter(|period| {
+                match Self::payout_period(&node, node.farm_id, node.connection_price, period) {
+                    Ok(_) => true,
+                    Err(e) => {
+                        log::debug!("payout failed: {:?}", e);
+                        false
+                    }
+                }
+            })
+            .collect();
+        PayablePeriods::<T>::insert(node_id, payable_periods);
+
+        Ok(().into())
+    }
+
+    pub fn payout_period(
+        node: &pallet_tfgrid::TfgridNode<T>,
+        farm_id: u32,
+        connection_price: u32,
+        period: &types::NodePeriodInformation,
+    ) -> DispatchResultWithPostInfo {
+        log::debug!("min capacity: {:?}", period.min_capacity);
+        let (cu, su) = period.min_capacity.cloud_units_permill();
+        log::debug!("cu: {}", cu);
+        log::debug!("su: {}", su);
+        let period_length = T::PeriodTreshold::get();
+
+        let farming_policy = pallet_tfgrid::FarmingPoliciesMap::<T>::get(period.farming_policy);
+        let farm = pallet_tfgrid::Farms::<T>::get(farm_id)
+            .ok_or(pallet_tfgrid::Error::<T>::FarmNotExists)?;
+        let farm_twin = pallet_tfgrid::Twins::<T>::get(farm.twin_id)
+            .ok_or(pallet_tfgrid::Error::<T>::TwinNotExists)?;
+
+        let mut cu_reward = 0;
+        let mut su_reward = 0;
+
+        let uptime_percentage = (period.uptime * 1_000 / period_length) as u128;
```

Comment:
not sure if `uptime <= period_length` is always guarantied
we should return some error if not the case